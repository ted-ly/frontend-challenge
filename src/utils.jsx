import { useState, useCallback } from 'react';

export const initialState = {
    name: '',
    password: '',
    color: '',
    email: '',
    terms: false,
};

export const onClickNext = (to, setPathname) => {
    setPathname(to);
    window.history.pushState(null, '', to);
};

export const onClickBack = (from, setPathname) => {
    setPathname(from);
    window.history.pushState(null, '', from);
}; 

export const onSubmit = async (formData, setPathname) => {
    const resp = await fetch('http://localhost:3001/api/submit', { method: 'POST', headers: {'Content-Type': 'application/json'}, body: JSON.stringify(formData) });
    if (resp.status >= 400) {
        window.history.pushState(null, '', '/error');
        setPathname('/error');
    } else {
        window.history.pushState(null, '', '/success');
        setPathname('/success');
    }
};

export const createOnClickRestart = ({ setFormData, setPathname }) => useCallback(() => {
    window.history.pushState(null, '', '/');
    setPathname('/');
    setFormData({...initialState});
}, []);

export const createOnFormDataChange = ({ setFormData, formData }) => useCallback((e) => {
    const name = e.target.name;
    const value = e.target.value;
    const newFormData = {
        ...formData,
        [name]: name ==='terms' ? !formData.terms : value,
    };

    setFormData({...newFormData});
}, [formData]); 

export const useForm = () => {
    const [formData, setFormData] = useState(initialState);
    const [pathname, setPathname] = useState('/');
    const onFormDataChange = createOnFormDataChange({ formData, setFormData });
    const onClickRestart = createOnClickRestart({ setFormData, setPathname });

    return { onClickRestart, onFormDataChange, formData, setPathname };
};

export const toCapitalized = (str) => {
    return str.slice(0, 1).toUpperCase() + str.slice(1);
}

export const toTransformText = (text, key) => {
    if (key === 'terms') {
        return text ? 'Agreed' : 'Disagreed';
    }

    if (key === 'password') {
        return '*'.repeat(text.length);
    }

    return text;
}; 