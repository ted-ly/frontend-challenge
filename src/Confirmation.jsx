import React, { useMemo, useCallback, useState, useEffect } from 'react';
import Button from './Button';
import Step from './Step';
import { onSubmit, toTransformText, onClickBack } from './utils';

export const textField = {
    name: 'First Name',
    email: 'E-mail',
    password: 'Password',
    color: 'Favorite Color',
    terms: 'Terms And Conditions',
}
export default function Confirmation({ formData, setPathname }) {
    const [isLoading, setIsLoading] = useState(false);
    const back = useCallback(() => {
        onClickBack('/more-info', setPathname);
    }, []);
    const confirmationData = useMemo(() => {
        return Object.keys(textField).map(key => <label className='confirmation-text' key={key} title={key}>{`${textField[key]}: ${toTransformText(formData[key], key)}`}</label>);
    }, []);

    const submit = useCallback(async (e) => {
        e.preventDefault();
        setIsLoading(true);
        await onSubmit(formData, setPathname);
        setIsLoading(false);
    }, []);
    return (
        <Step title='Confirmation' isLoading={isLoading} to='confirmation'>
            <div className='confirmation-container'>
                {confirmationData}
                <div>
                    <Button onClick={back}>Back</Button>
                    <Button onClick={submit} className='next-btn'>Submit</Button>
                </div>
            </div>
        </Step>
    );
}