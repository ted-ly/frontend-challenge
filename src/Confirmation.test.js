import renderer from 'react-test-renderer';
import Confirmation from './Confirmation';
import { initialState } from './utils';

describe('AdditionalInfo', () => {
    it('should match snapshot', () => {
        const component = renderer.create(<Confirmation formData={initialState} />);
        expect(component.toJSON()).toMatchSnapshot();
    });
});