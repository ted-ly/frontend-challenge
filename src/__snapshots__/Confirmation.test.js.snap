// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdditionalInfo should match snapshot 1`] = `
<div
  className="step-container"
>
  <div
    className="step-title"
  >
    Confirmation
  </div>
  <div
    className="confirmation-container"
  >
    <label
      className="confirmation-text"
      title="name"
    >
      First Name: 
    </label>
    <label
      className="confirmation-text"
      title="email"
    >
      E-mail: 
    </label>
    <label
      className="confirmation-text"
      title="password"
    >
      Password: 
    </label>
    <label
      className="confirmation-text"
      title="color"
    >
      Favorite Color: 
    </label>
    <label
      className="confirmation-text"
      title="terms"
    >
      Terms And Conditions: Disagreed
    </label>
    <div>
      <button
        onClick={[Function]}
        type="button"
      >
        Back
      </button>
      <button
        className="next-btn"
        onClick={[Function]}
        type="button"
      >
        Submit
      </button>
    </div>
  </div>
</div>
`;
